app/gen-server/sqlUtils.ts

```diff
@@ -125,6 +127,37 @@ export function datetime(dbType: DatabaseType) {
   }
 }
 
+export function addParamToQuery(dbType: DatabaseType, index: number) {
+  switch (dbType) {
+    case 'postgres':
+      return `$${index}`;
+    case 'sqlite':
+      return `?`;
+    default:
+      throw new Error(`addParamToQuery not implemented for ${dbType}`);
+  }
+}
+
+export async function addRefToUserList(queryRunner: QueryRunner, userList: any[]){
+  const dbType = queryRunner.connection.driver.options.type;
+
+  // Updating so many rows in a multiple queries is not ideal. We will send updates in chunks.
+  // 300 seems to be a good number, for 24k rows we have 80 queries.
+  const userChunks = chunk(userList, 300);
+  for (const users of userChunks) {
+    await queryRunner.connection.transaction(async manager => {
+      const queries = users.map((user: any, _index: number, _array: any[]) => {
+        return manager.query(
+          `UPDATE users
+          SET ref = ${addParamToQuery(dbType, 1)}
```

Comment:
Overall with the `addParamToQuery` function included this feels quite messy compared to the original `queryRunner.manager.save` setup? I probably missed it earlier in development, but what was the main reason for the change?

Reply:
There's some context [here](https://github.com/gristlabs/grist-core/pull/935#discussion_r1565835770).

Reply:
Thanks, I understand now. Would it be possible to narrow the scope of the selects, like for example in https://github.com/gristlabs/grist-core/blob/5956c20c295f3f6f94ca1a7941a28ae3a091c0e8/app/gen-server/migration/1568238234987-TeamMembers.ts#L10-L33? Hmm but then there is the actual updates to do...

Bundling does seem important since there are a lot of users. But I'm afraid if this code had a subtle bug we wouldn't catch it, and previous times we have had subtle bugs in migrations it has been a mess. I think I need to ask for this method's functionality to be given a set of unit tests. The addParamToQuery function that does one thing for Postgres and then ignores an argument for SQLite, relying on the user to call things in the right order, also feels dangerous - does it really need to be exported, can it be at least kept private?

Reply:
In my last commit I added a test to be sure it enters inside the for loop of the `addRefToUserList` function and I remove the `addParamToQuery` function that is not safe